### DISCUSSION TITLE
Elsa Workflow Meeting #29

### DISCUSSION THREAD

**Discussion by gentledepp** [elsa-workflows/elsa-core#3852] [Ideas]
Hello! I recently started watching your meetings and, having integrated another WF library in the passt, there were several thoughts that came to my mind:

1.) CorrelationId
We are developing a sofotware that uses lists similar to SharePoint, where there is a task list and every time a new item is added, we want to run a workflow for this task.
Now, based on what was said in the video, one has to explicitly model this using a Correlate activity in a workflow.
However, this may be a bit too much to ask from our not-so-tech-savvy user.
So is it also possible to set the correlationid in a workflow context provider?

2.) "Randomness" of start activities
It seems to me like elsa is often guessing which activities it uses as start activites in a wf.
I thing this is too error-prone and there should simply be the possibility to mark activities as start- and end-activities. Of course the designer can flag any activity that has no predecessor by default.
But it should be possible to override this and there should be a solid validation in the designer to ensure that a WF always has at lease one start- and end-activity.
What do you think?

3.) Typescript support
When working with custom data, such as an HTTP Post coming in, in JS this is a simple object, and in C# an ExpandoObject. So without any type info, it is quite easy to add typos in JavaScript activities.
My whish would be that it was possible to define a typescript interface for the incoming JSON.
Then, the monaco editors could give us hints in case we typed some property wrong or if we did not check for an optional argument to exist.
Maybe this could even be used to create a swagger UI definition so it is easy to start this WF as an API developer. 😁

4.) Elas SSAS
For your Elas-as-a-service idea, I recently came across Steve Sandersons DotNetIsolator:
https://github.com/SteveSandersonMS/DotNetIsolator
It allows you to run any C# assembly in a WASM sandbox.
Wouldn't this be _the_ solution for Elas SAAS? Developers could still add custom activities but they would be safely executed in a lightweight sandbox! 🦾

I would love to hear/read your feedback on those 4 topics @sfmskywalker

**Comment by sfmskywalker** [comment]
Hi @gentledepp , thank you for your valuable feedback!

1. CorrelationId
Yes, the Set Correlation Id activity is just one way to correlate the workflow. Another way is to include a correlation ID with e.g. HTTP requests that trigger new workflows, or when you invoke a workflow programmatically, you can supply a correlation ID there as well.
And we'll add more ways to do this based on any future feedback.

2. "Randomness" of start activities
I agree, there should be a way to explicitly mark the starting activity so that there is no ambiguity. 

3. Typescript support
It makes sense. Elsa 2 solves this by letting the user define a JSON schema, which is used by the intellisense engine as well. We should do the same for Elsa 3, and perhaps even also add the option to define the schema in TypeScript as well.

4. Elsa SaaS
I have to dig deeper into the DotNetIsolator project, but from first looks, it does look perfect for programatically spinning up isolated instances of Elsa and, indeed, allow the user to install assemblies isolated to their own instance!
I bookmarked this repo for future exploration.

Thanks again for the great input!

**Reply by gentledepp** [reply]
You mentioned the JSON schema support - so I started digging a bit.
Since OpenAPI uses a variant of JSON schema, I thought: Why can't we 
* define the input model as a JSON schema **or** a typescript interface
* and even use this to power Swagger...

and the result:
There is a MIT-licensed project "tsoa" which does just that.
It allows to define APIs in typescript using node and express and creates the OpenAPI documentation of it:
https://github.com/lukeautry/tsoa

So in my opinion it would be awesome if:
- we could define the input model using JSON Schema **or** typescript interface
- and this could then power the monaco editor for auto-complete and syntax checking
- and, _if_ the workflow can be triggered via http, we could even create a swagger doc for it.

Does this make sense?
Would you say this'd be useful and doable?

**Reply by sfmskywalker** [reply]
Absolutely useful and doable. In fact, Elsa 2 does something similar (if not the same). There, the HTTP Endpoint activity also allows the user to enter a JSON schema, which is then used server-side to generate a type definition, used to power the intellisense feature of the Monaco editor. The implementation is a bit messy, but we could do something elegant in V3.

As for Swagger, that works (or used to work) in V2 for HTTP workflows, and we should add it to V3 as well.

**Comment by GeorgeS2019** [comment]
@gentledepp 

WASM using Blazor
https://github.com/elsa-workflows/elsa-core/issues/3232#issuecomment-1508155158
https://github.com/elsa-workflows/elsa-core/issues/3810
